includes/utils/class-amp-image-dimension-extractor.php

```diff
@@ -47,73 +71,157 @@ public static function normalize_url( $url ) {
-			return false;
-		}
-		set_transient( $transient_lock_name, 1, MINUTE_IN_SECONDS );
-
-		// Note to other developers: please don't use this class directly as it may not stick around forever...
-		if ( ! class_exists( 'FastImage' ) ) {
-			require_once( AMP__DIR__ . '/includes/lib/class-fastimage.php' );
-		}
-
-		// TODO: look into using curl+stream (https://github.com/willwashburn/FasterImage)
-		$image = new FastImage( $url );
-		$dimensions = $image->getSize();
-
-		if ( ! is_array( $dimensions ) ) {
-			set_transient( $transient_name, $transient_fail, $transient_expiry );
-			delete_transient( $transient_lock_name );
-			return false;
-		}
-
-		set_transient( $transient_name, $dimensions, $transient_expiry );
-		delete_transient( $transient_lock_name );
-		return $dimensions;
-	}
+    /**
+     * Determine which images to fetch by checking for dimensions in transient/cache.
+     * Creates a short lived transient that acts as a semaphore so that another visitor
+     * doesn't trigger a remote fetch for the same image at the same time.
+     *
+     * @param array $urls Urls of image src
+     * @param array $urls_to_fetch Urls of images to fetch because dimensions are not in transient/cache
+     * @param array $all_dimensions "Master" list of img url to dimension mappings - used here to track which dimensions couldn't be retrieved
+     */
+	private static function determine_which_images_to_fetch(&$urls, &$urls_to_fetch, &$all_dimensions) {
+        foreach ( $urls as $url ) {
+            $url_hash = md5( $url );
+            $transient_name = sprintf( 'amp_img_%s', $url_hash );
+            $dimensions = get_transient( $transient_name );
+
+            if ( is_array( $dimensions ) ) {
+                $all_dimensions[$url] = array(
+                    'width' => $dimensions[0],
+                    'height' => $dimensions[1],
+                );
+            } else {
+                if ( self::STATUS_FAILED_LAST_ATTEMPT === $dimensions ) {
+                    // If we couldn't get dimensions for this image the last time we tried, don't try again until the transient expires
+                    $all_dimensions[$url] = false;
+                    continue;
+                } else {
+                    // Ensure we're the only one trying to extract dimensions for this image at the moment so we don't duplicate requests
+                    $transient_lock_name = sprintf( 'amp_lock_%s', $url_hash );
+                    if ( false !== get_transient( $transient_lock_name ) ) {
+                        $all_dimensions[$url] = false;
+                        continue;
+                    } else {
+                        $urls_to_fetch[$url]['url'] = $url;
+                        $urls_to_fetch[$url]['transient_name'] = $transient_name;
+                        $urls_to_fetch[$url]['transient_lock_name'] = $transient_lock_name;
+                        set_transient($transient_lock_name, 1, MINUTE_IN_SECONDS);
```

Comment:
We're getting pretty deeply nested here. Any way we can flatten the conditional blocks a bit?